src/eventbuffer/InMemoryJSONEventBuffer.ts

```diff
@@ -488,46 +435,72 @@ export default class InMemoryJSONEventBuffer implements EventBuffer<EventData>,
           this.logger.debug(
             `Event Reporting - InMemoryJSONEventBuffer - sendEventImmediately - beaconing data out ${body}`
           );
-          /* istanbul ignore else */
           if (!navigator.sendBeacon(`${this.ingestionURL}?beacon=1`, body)) {
             failed = true;
+          } else {
+            failed = false;
           }
         }
       } catch (error) {
-        this.logger.warn(`Event Reporting - Error sending beacon for an important event ${name}`);
+        this.logger.warn(`Event Reporting - Error sending beacon for an important event ${body}`);
         failed = true;
       }
     }
 
     /* istanbul ignore else */
     if (failed) {
       this.logger.debug(
-        `Event Reporting - InMemoryJSONEventBuffer - sendEventImmediately - pushing to failed events ${JSON.stringify(
-          event
-        )}`
+        `Event Reporting - InMemoryJSONEventBuffer - sendEventImmediately - pushing to failed events ${body}`
       );
       this.failedIngestionEvents.push(event);
     }
   }
 
+  private getBackoffWaitTime(retryCount: number): number {
+    return (
+      Math.random() *
+      Math.min(
+        InMemoryJSONEventBuffer.RETRY_LONG_BACKOFF_MS,
+        InMemoryJSONEventBuffer.RETRY_SHORT_BACKOFF_MS * Math.pow(2.0, retryCount)
+      )
+    );
+  }
+
+  private wait(delayInMs: number): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, delayInMs));
+  }
+
   private async send(data: string): Promise<Response> {
     try {
-      const response = await fetch(this.ingestionURL, {
-        method: 'POST',
-        headers: {
-          Authorization: `Bearer ${this.authenticationToken}`,
-        },
-        body: data,
-      });
-      return response;
+      let retryCount = 0;
+      while (retryCount < this.retryCountLimit) {
+        const response = await fetch(this.ingestionURL, {
+          method: 'POST',
+          headers: {
+            Authorization: `Bearer ${this.authenticationToken}`,
+          },
+          body: data,
+        });
+        if (response.ok || !InMemoryJSONEventBuffer.SENDING_FAILURE_CODES.has(response.status)) {
+          return response;
+        } else {
+          retryCount++;
+          const backoffTime = this.getBackoffWaitTime(retryCount);
```

Comment:
You can skip waiting step if `retryCount == this.retryCountLimit`

Reply:
Makes sense, adding a check here. Nice call!